src/representation_theory/monoid_algebra_basis.lean

```diff
@@ -0,0 +1,189 @@
+/-
+Copyright (c) 2022 Amelia Livingston. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Amelia Livingston
+-/
+
+import representation_theory.Rep
+import representation_theory.basic
+
+/-!
+# The structure of the `k[G]`-module `k[Gⁿ]`
+
+This file contains API for the `k[G]`-module `k[Gⁿ]`, where `k` is a commutative ring, `G` is
+a group, and we express the module structure as the representation `G →* End(k[Gⁿ])` induced by the
+diagonal action of `G` on `Gⁿ.`
+
+In particular, we define morphisms of `k`-linear `G`-representations between `k[Gⁿ⁺¹]` and
+`k[G] ⊗ₖ k[Gⁿ]` (on which `G` acts by `ρ(g₁)(g₂ ⊗ x) = (g₁ * g₂) ⊗ x`) which we will later show
+define an isomorphism. From this we will deduce that `k[Gⁿ⁺¹]` is free over `k[G]`.
+The freeness of these modules means we can use them to build a projective resolution of the
+(trivial) `k[G]`-module `k`, which is useful in group cohomology.
+
+## Main definitions
+
+  * representation.of_mul_action
+  * Rep.of_mul_action
+  * to_tensor
+  * of_tensor
+
+## Implementation notes
+
+We express `k[G]`-module structures on a group `V` as `k`-linear representations of `G` on `V`, as
+this simplifies some proofs and also makes it easier to work with different structures on the same
+`V`. This is because `module` is a class, and there is only one notation for scalar multiplication,
+`•`; representations, meanwhile, give us more flexibility, by avoiding typeclass inference.
+
+We bundle the type `k[Gⁿ]` - or more generally `k[H]` when `H` has `G`-action - and the
+representation together as a term of type `Rep k G`, and we call it `Rep.of_mul_action k G H.` This
+is so we can talk about morphisms of representations.
+
+-/
+
+noncomputable theory
+universes v u
+variables (k G : Type u) [comm_ring k]
+
+open monoid_algebra
+
+namespace representation
+variables [monoid G] (H : Type u) [mul_action G H]
+
+/-- A `G`-action on `H` induces a representation `G →* End(k[H])` in the natural way. -/
+@[simps] def of_mul_action : representation k G (H →₀ k) :=
+{ to_fun := λ g, finsupp.lmap_domain k k ((•) g),
+  map_one' := by { ext x y, dsimp, simp },
+  map_mul' := λ x y, by { ext z w, simp [mul_smul] }}
+
```

Comment:
I think it's worth adding:
```lean
lemma of_mul_action_apply' {G H : Type u} [group G] [mul_action G H]
  (g : G) (f : H →₀ k) (h : H) : of_mul_action k G H g f h = f (g⁻¹ • h) :=
begin
  conv_lhs { rw ← smul_inv_smul g h, },
  let h' := g⁻¹ • h,
  change of_mul_action k G H g f (g • h') = f h',
  have hg : function.injective ((•) g : H → H), { intros h₁ h₂, simp, },
  simp only [of_mul_action_apply, finsupp.lmap_domain_apply, finsupp.map_domain_apply, hg],
end
```
(it might even be giving this the `simp` attribute rather than `of_mul_action_apply` generated by `simps` since I imagine we'll often have `group G` rather than `monoid G`.)

Reply:
Thank you, I have done this & also written the `simps` lemma explicitly (without a `simp` tag) as `of_mul_action_def`.